oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,438 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! HTTP server pseudo-Node that can serve as a frontend for an Oak node.
+//! The server receives requests from the outside, wraps each request in
+//! an invocation and sends it to the designated Oak node to be processed
+//! asynchronously.
+
+use crate::{
+    node::{ConfigurationError, Node},
+    runtime::{ChannelHalfDirection, RuntimeProxy},
+};
+use hyper::{
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server, StatusCode,
+};
+use log::{debug, error, info, warn};
+use oak_abi::{
+    label::Label,
+    proto::oak::{
+        application::HttpServerConfiguration,
+        encap::{HttpRequest, HttpResponse},
+    },
+    ChannelReadStatus, OakStatus,
+};
+use prost::Message;
+use std::{collections::HashMap, future::Future, net::SocketAddr, pin::Pin};
+use tokio::sync::oneshot;
+
+#[cfg(test)]
+pub mod tests;
+
+/// Struct that represents an HTTP server pseudo-Node.
+pub struct HttpServerNode {
```

Comment:
@daviddrysdale I am still working on this, but could you please take a look and let me know if you think I am going the wrong direction? :) 
It is mostly copied from `GrpcServerNode`, but I am trying to replace the `tonic` stuff with `hyper`.

Reply:
Seems plausible so far!  Couple of early suggestions:
 - It seems like you're encapsulating things into separate functions (`try_get_invocation_channel`) and types (`Pipe`) a bit more than the existing gRPC server code does.  That's a good idea, so I'd suggest applying it to the existing `node/grpc/server/mod.rs` code as an initial preparatory PR, making sure to put the extracted code somewhere (`node/utils.rs` maybe) that this PR can just `use` later (rather than having 2 copies).
 - I have a suspicion that the `async` stuff can be simpler here, because we're not going to need to support anything equivalent to server-streaming responses (I assume).  Maybe look at the gRPC server code pre-#1117 and see if there's a simpler structure that would still work?

Reply:
Thanks for the comments @daviddrysdale :) I am also a bit unsure about server-streaming. I'll look at `GrpcServerNode` before #1117, and see if things can be simplified here. 

I agree with basing `GrpcServerNode` and `HttpServerNode` on the same abstractions. I think I have a clearer idea what that would be after working on this PR, and your comment below. It might still be easier for me to extract out the similar functionality after this PR (particularly, to have tests for both GRPC and HTTP server p-Ns). In any case, I'll coordinate any changes to the `GrpcServerNode` with @conradgrobler and @ipetr0v.

Reply:
Just following up &ndash; did any commonization of the code between gRPC and HTTP server pseudo-Nodes happen?

Reply:
Not yet. Tiziano and I have briefly discussed it. I am adding a minimal SDK support for HTTP server p-N to be able to use it in examples, and write more tests. Then I'll do the refactoring.

Reply:
@daviddrysdale Regarding commonization, I have some clearer ideas now, and I have created #1390 to track that.